src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,468 @@
+//! This module defines the shapes and logic required for drawing node ports.
+
+use crate::prelude::*;
+
+use crate::component::node::Node;
+
+use core::f32::consts::PI;
+use ensogl::data::color::*;
+use ensogl::display::Attribute;
+use ensogl::display::Buffer;
+use ensogl::display::Scene;
+use ensogl::display::object::Object;
+use ensogl::display::object::ObjectOps;
+use ensogl::display::scene::ShapeRegistry;
+use ensogl::display::shape::*;
+use ensogl::display::shape::primitive::def::class::ShapeOps;
+use ensogl::display::symbol::geometry::Sprite;
+use ensogl::display;
+use ensogl::gui::component::ShapeViewDefinition;
+use ensogl::gui::component;
+use ensogl::math::geometry::circle::segment::Segment;
+use ensogl::math::geometry::triangle;
+use ensogl::math::topology::unit::Angle;
+use ensogl::math::topology::unit::AngleOps;
+use ensogl::math::topology::unit::Degrees;
+use ensogl::math::topology::unit::Distance;
+use ensogl::math::topology::unit::Pixels;
+
+
+
+// ===========================
+// === Port Specification ===
+// ===========================
+
+/// Indicates whether a port is inwards facing  or outwards facing.
+#[allow(missing_docs)]
+#[derive(Clone,Copy,Debug)]
+pub enum Direction{
+    In,
+    Out,
+}
+
+/// Defines the properties of a port shape and can then
+/// be used to build the port shape.
+///
+/// Ports are constructed around an inner circle, and thus
+/// most measurements are in degrees, which are measured around
+/// a inner circle that is defined by the `inner_radius`.
+#[derive(Clone,Copy,Debug)]
+pub struct Specification {
+    /// Height of the port.
+    pub height      : f32,
+    /// Width of the port in degrees.
+    pub width       : Angle<Degrees>,
+    /// Radius of the inner circle that the port is constructed around.
+    pub inner_radius: f32,
+    /// Location of the port along the inner circle.
+    pub position: Angle<Degrees>,
```

Comment:
I would name it rotation instead.

Reply:
I think that could lead to confusion about what is meant here since the rotation could also just men the "orientation" of the object. I think a position expressed as `Angle` might be clearer.